### PR TITLE
fix(serializer): preserve Decimal values and non-primitive dict keys

### PR DESCRIPTION
**Pull request by i-anubhav-anand**
Two cases where `EventSerializer` silently dropped data in serialized trace input/output:

### 1. `decimal.Decimal` values are lost
There is no `Decimal` branch in `_default_inner`, so a `Decimal` falls through to the `"<{type}>"` fallback:
```python
EventSerializer().encode(Decimal("1.5"))            # -> "\"<Decimal>\""
EventSerializer().encode({"price": Decimal("19.99")})  # -> {"price": "<Decimal>"}
```
`Decimal` is common in financial/tool outputs. Fix serializes it via `str()` so the exact value is preserved (`float()` would silently round high-precision values and can overflow on very large ones); `Decimal("NaN")`/`Decimal("Infinity")` still render as `"NaN"`/`"Infinity"`.

### 2. A non-primitive dict key discards the whole dict
The dict branch did `{self.default(k): self.default(v) ...}`. When a key serializes to a container/object (tuple, set, custom object), the comprehension/JSON encoding raises and the `except` turns the **entire dict** into an error string:
```python
EventSerializer().encode({(1, 2): "v"})  # -> "\"<not serializable object of type: dict>\""
```
Fix stringifies such keys via `str(key)` so a single non-primitive key no longer drops every value.

### Tests
Adds 4 unit tests in `tests/unit/test_serializer.py` (verified fail-before / pass-after). Existing `datetime`/`uuid`/`enum`/`int` key behavior is unchanged (existing parametrized test still passes). `ruff check`/`ruff format` clean.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes two silent data-loss bugs in `EventSerializer`: `decimal.Decimal` values were falling through to the `"<Decimal>"` opaque fallback, and a dict with any non-primitive key would discard the entire dict instead of just stringifying that key.

- **Decimal fix**: `Decimal` is now routed through `float()` before the existing `NaN`/`Infinity` handlers, so it serializes as a JSON number. Special `Decimal` values (`NaN`, `Infinity`, `-Infinity`) are handled correctly. The precision trade-off (float has ~15–17 significant digits) is an intentional design choice but is undocumented.
- **Dict key fix**: The dict branch now checks whether a serialized key is a JSON-scalar type, and falls back to `str(k)` for non-scalars (tuples, sets, custom objects), so a single exotic key no longer discards every value in the dict.
- Four new unit tests verify both fixes, including special Decimal values, tuple keys, and custom-object keys.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge — both fixes address real silent-data-loss regressions, the logic is correct, and the new tests exercise the changed paths including edge cases.

The Decimal-to-float conversion works correctly for normal and special values, but silently narrows high-precision Decimal values to float precision without any documentation or warning. For a tracing library this is unlikely to cause problems in practice, but users who pass high-precision financial Decimals would see quietly rounded values in their trace data.

The Decimal precision trade-off in langfuse/_utils/serializer.py (lines 79–80) is the only area worth a second look; tests/unit/test_serializer.py is straightforward.
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
langfuse/_utils/serializer.py:79-80
**Decimal precision silently lost on float conversion**

`float()` can only represent ~15–17 significant digits, so any `Decimal` with higher precision is silently rounded. For example, `Decimal("1.0000000000000001")` becomes `1.0` and `Decimal("123456789012345678")` would gain rounding error. Since `Decimal` is often chosen specifically to avoid float imprecision (financial, scientific), users may find their high-precision values quietly corrupted in trace data. A comment documenting this trade-off would help, or alternatively the value could be converted to `str` instead, preserving the exact decimal representation while still avoiding the opaque `"<Decimal>"` fallback.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(serializer): preserve Decimal values..."](https://github.com/langfuse/langfuse-python/commit/ad99074df1c494060ebd8b4af8e6d38774342cae) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37503538)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: langfuse/_utils/serializer.py [hunk 1]
```diff
@@ -72,6 +73,12 @@ def _default_inner(self, obj: Any) -> Any:
             if np is not None and isinstance(obj, np.ndarray):
                 return obj.tolist()
 
+            # Route Decimal through float so it is preserved as a number
+            # (and Decimal("NaN")/Decimal("Infinity") reuse the handlers below)
+            # instead of falling through to the "<Decimal>" type fallback.
+            if isinstance(obj, decimal.Decimal):
+                return self.default(float(obj))
```
[the remark this answers is not in the document]
Comment (i-anubhav-anand) [reply langfuse/_utils/serializer.py]: Addressed in the latest commit — switched to `str(obj)` so the exact Decimal value is preserved (this also avoids the OverflowError that `float()` raises on very large Decimals). NaN/Infinity still render as "NaN"/"Infinity". Tests updated to assert exact string output incl. high-precision and >JS-safe-int cases.
